torch/utils/data/common.py

```diff
@@ -0,0 +1,162 @@
+import sys
+import os
+import fnmatch
+import warnings
+import tarfile
+import zipfile
+from typing import List, Iterable, Union, Any
+from io import IOBase
+
+def match_masks(name : str, masks : Union[str, List[str]]):
+    # empty mask matches any input name
+    if not masks:
+        return True
+
+    if isinstance(masks, str):
+        return fnmatch.fnmatch(name, masks)
+
+    for mask in masks:
+        if fnmatch.fnmatch(name, mask):
+            return True
+    return False
+
+def get_file_pathnames_from_root(
+        root: str,
+        masks: Union[str, List[str]],
+        recursive: bool = False,
+        abspath: bool = False) :
+
+    def onerror(err : OSError):
+        warnings.warn(err.strerror + " : " + err.filename)
+
+    for path, dirs, files in os.walk(root, onerror=onerror):
+        if abspath:
+            path = os.path.abspath(path)
+        for f in files:
+            if match_masks(f, masks):
+                yield os.path.join(path, f)
+        if not recursive:
+            break
+
+
+def extract_files_from_single_pathname_binary(
+        pathname : str,
+        binary_stream : Any,
+        extract_tars : bool = True,
+        extract_zips : bool = True,
+        recursive : bool = True,
+        mask_tars : Union[str, List[str]] = '',
+        mask_zips : Union[str, List[str]] = ''):
+
+    # test whether binary_stream is seekable (eg. PIPE stream from webdata is not seekable)
+    seekable = hasattr(binary_stream, "seekable") and binary_stream.seekable()
+
+    # Note: We have no way to verify whether a non-seekable stream (eg. PIPE stream) is tar/zip without
+    #       changing stream handle position, however, there is no way to move such stream's handle back.
+    #       For non-seekable stream, the entire stream will be discard if it can not be extracted
+
+    # extract tar file
+    if extract_tars and match_masks(pathname, mask_tars):
+        try:
+            with tarfile.open(fileobj=binary_stream) as tar:
+                for tarinfo in tar:
+                    if not tarinfo.isfile():
+                        continue
+
+                    extract_fobj = tar.extractfile(tarinfo)
+                    if extract_fobj is None:
+                        warnings.warn("failed to extract tar file {}".format(tarinfo.name))
+                        continue
+
+                    inner_pathname = os.path.normpath(os.path.join(pathname, tarinfo.name))
+                    if recursive:
+                        yield from extract_files_from_single_pathname_binary(
+                            inner_pathname, extract_fobj, extract_tars, extract_zips, recursive, mask_tars, mask_zips)
+                    else:
+                        yield (inner_pathname, extract_fobj)
+            return
+        except tarfile.TarError:
+            if not seekable:
```

Comment:
Please raise Exception in this case, we don't want to silently swallow errors in care of bad files